### DISCUSSION TITLE
[TypeScript] access parameters/responses subtypes

### DISCUSSION THREAD

**Discussion by ianwremmel** [octokit/octokit.js#1697] [General]
I've been making use of the types `AppsListAccountsUserOrOrgOnPlanResponse`, `AppsListAccountsUserOrOrgOnPlanResponseItem`, and `AppsListAccountsUserOrOrgOnPlanResponseItemMarketplacePurchasePlan`.

With V17, I can do 
```ts
export type AppsListAccountsUserOrOrgOnPlanResponse = RestEndpointMethodTypes['apps']['listAccountsUserOrOrgOnPlan']['response']['data'];
```
But to get to the subtypes, I need to do 

```ts
export type AppsListAccountsUserOrOrgOnPlanResponse = RestEndpointMethodTypes['apps']['listAccountsUserOrOrgOnPlan']['response']['data'];
export type AppsListAccountsUserOrOrgOnPlanResponseItem = AppsListAccountsUserOrOrgOnPlanResponse[0];
export type AppsListAccountsUserOrOrgOnPlanResponseItemMarketplacePurchasePlan = AppsListAccountsUserOrOrgOnPlanResponseItem['marketplace_purchase']['plan']
```

In articular, the `AppsListAccountsUserOrOrgOnPlanResponse[0]` seems strange.

**Comment by gr2m** [comment]
Sorry Ian for not getting back to you. My time for Octokit is very limited, and I don't have a good answer. But I do agree that the current solution is not sufficient.

I am considering to export the types similar to how they have been exported in v16. I'm not quite sure what the right approach would be, maybe export a`Octokit` namespace with an `Endpoints` interface on it, which would then have all the different keys for requests and responses on it? 

Can you think of a project similar to Octokit which does a good job with that? Or anyone I could reach out to for help?

**Reply by ianwremmel** [reply]
> Sorry Ian for not getting back to you.

Well, as long as you won't hold the same against me :) 

> Can you think of a project similar to Octokit which does a good job with that?

The only other project I'm aware that would expect to use a typescript setup like octokit is the AWS SDK. I don't know if it's generated in a similar way, but I do believe quite a lot if its code is generated.

**Comment by robsonmeemo** [comment]
It gets uglier with optional types:
```
type ChecksCreateParamsOutputAnnotations = NonNullable<
  NonNullable<
    RestEndpointMethodTypes["checks"]["create"]["parameters"]["output"]
  >["annotations"]
>[0];
```

Is it feasible to just have aliases for those types?

**Reply by hipstersmoothie** [reply]
I've had to do this too. The `NonNullable` makes it very ugly.

**Comment by gr2m** [comment]
I have a question: say I do export all the types and sub types for the endpoint parameters and responses, such as `AppsListAccountsUserOrOrgOnPlanResponseItemMarketplacePurchasePlan`

What happens frequently is that the endpoint names are changing. Now `octokit.apps.listAccountsUserOrOrgOnPlan` got renamed `octokit.apps.listAccountsForPlan`. For the methods, I keep track of these renames, I keep both methods working and log a deprecation message for the former. With v18, I plan to only remove previously deprecated messages.

For TypeScript, I don't know how to communicate that a type was deprecated. Is there such a possibility? I think it's possible for methods, but not types? Any thoughts on that?

**Reply by robsonmeemo** [reply]
How would it work today with the renames?

**Reply by gr2m** [reply]
It wouldn't work with `RestEndpointMethodTypes`. But that's why I was hoping that extracting the types from the methods using `GetResponseTypeFromEndpointMethod` and `GetResponseDataTypeFromEndpointMethod` would be an alternative, see https://github.com/octokit/types.ts/#get-response-types-from-endpoint-methods

THe other alternative is to get the types by REST API path:
https://github.com/octokit/types.ts/#get-parameter-and-response-data-types-for-a-rest-api-endpoint

But then again, accessing subtypes is cumbersome for both.

The alternative is to support both `AppsListAccountsUserOrOrgOnPlanResponseItemMarketplacePurchasePlan` and `AppsListAccountsForPlanResponseItemMarketplacePurchasePlan`, without the convenience of a deprecation message as there is for the actual methods

**Reply by ianwremmel** [reply]
As far as renaming things, it seems like an `@deprecated` jsdoc is the preferred way, even thought it's not quite standard yet.

- https://github.com/microsoft/TypeScript/issues/390
- https://github.com/microsoft/vscode/issues/53514

**Reply by ianwremmel** [reply]
Oh, I see, you were concerned about deprecating interfaces. I'm not sure if `@deprecated` works there or not.

**Reply by ianwremmel** [reply]
maybe it's just a matter of maintaining all names for some amount of time, and when it reaches an annoying level of maintenance, put out a major update?

**Reply by gr2m** [reply]
> when it reaches an annoying level of maintenance, put out a major update

That is basically how I do versions today already. I avoid breaking changes at all costs, but eventually bump the major version to get rid of all the deprecations. As part of the migration to the new version, I ask folks to update to the latest version of the previous major version and address all deprecations they see in their tests. 

But I think there are no deprecation messages for TypeScript interfaces, so people will have to look up the new names manually. Not ideal, but also not the end of the world :)

**Reply by ecraig12345** [reply]
Coming in late here, but I'd personally prefer to have explicit interfaces even if I sometimes have to update the names after major releases. If the interface names correspond to method names, and the methods *do* log deprecation warnings at runtime, that would hopefully help clue people in to update the interface names in their code at the same as they update the method names.

Another way to get type deprecation warnings is that if you add `@deprecated` to the JSDoc, [`eslint-plugin-deprecation`](https://www.npmjs.com/package/eslint-plugin-deprecation) (or tslint's `deprecation` rule) can be used to get lint errors about usage of deprecated things, though of course not everyone has that set up.
